src/Signer.php

```diff
@@ -13,45 +13,69 @@
 
 namespace Nelmio\SecurityBundle;
 
-final class Signer
+use Nelmio\SecurityBundle\Signer\SignerInterface;
+
+final class Signer implements SignerInterface
 {
     private string $secret;
     private string $algo;
+    private ?string $legacyAlgo;
+
+    /**
+     * @var non-empty-string
+     */
+    private string $separator;
 
-    public function __construct(string $secret, string $algo)
+    /**
+     * @param non-empty-string $separator
+     */
+    public function __construct(string $secret, string $algo, ?string $legacyAlgo = null, string $separator = '.')
     {
         $this->secret = $secret;
         $this->algo = $algo;
+        $this->legacyAlgo = $legacyAlgo;
+        $this->separator = $separator;
 
         if (!\in_array($this->algo, hash_algos(), true)) {
             throw new \InvalidArgumentException(sprintf("The supplied hashing algorithm '%s' is not supported by this system.", $this->algo));
         }
+
+        if (null !== $this->legacyAlgo && !\in_array($this->legacyAlgo, hash_algos(), true)) {
+            throw new \InvalidArgumentException(sprintf("The supplied legacy hashing algorithm '%s' is not supported by this system.", $this->legacyAlgo));
+        }
     }
 
     public function getSignedValue(string $value, ?string $signature = null): string
     {
         if (null === $signature) {
-            $signature = $this->generateSignature($value);
+            $signature = $this->generateSignature($value, $this->algo);
         }
 
-        return $value.'.'.$signature;
+        return $value.$this->separator.$signature;
     }
 
     public function verifySignedValue(string $signedValue): bool
     {
         [$value, $signature] = $this->splitSignatureFromSignedValue($signedValue);
-        $signature2 = $this->generateSignature($value);
+        if (null === $signature) {
+            return false;
+        }
+
+        $expectedSignature = $this->generateSignature($value, $this->algo);
+        if (hash_equals($expectedSignature, $signature)) {
+            return true;
+        }
 
-        if (null === $signature || \strlen($signature) !== \strlen($signature2)) {
+        if (null === $this->legacyAlgo) {
             return false;
         }
 
-        $result = 0;
-        for ($i = 0, $j = \strlen($signature); $i < $j; ++$i) {
-            $result |= \ord($signature[$i]) ^ \ord($signature2[$i]);
+        $expectedLegacySignature = $this->generateSignature($value, $this->legacyAlgo);
+        if (hash_equals($expectedLegacySignature, $signature)) {
```

Comment:
Ah nice, didn't realize we still were using some legacy version of hash_equals here :D